torch_ttnn/passes/lowering/to_tt_pass.py

```diff
@@ -638,6 +638,31 @@ def rewrite_node(node):
                     input = g.call_function(ttnn.to_layout, args=(input, TtnnRowMajorLayout()))
                 return g.call_function(ttnn.pad, args=(input, full_pad, value))
 
+            # NOTE(jdh8): We only convert max_pool2d without indices
+            if node.target == torch.ops.aten.max_pool2d_with_indices.default:
+                try:
+                    (user,) = node.users.keys()
+                    assert user.target.__name__ == "getitem"
+                    assert user.args == (node, 0)
+                except:
+                    return None
+
+                in_n, in_c, in_h, in_w = args[0].meta["val"].size()
+                out_n, out_c, out_h, out_w = node.meta["val"][0].size()
+                input_nhwc = g.call_function(ttnn.permute, (args[0], (0, 2, 3, 1)))
+                input_nhwc = g.call_function(ttnn.reshape, (input_nhwc, (-1, in_c)))
+                kwargs = {
+                    "stride": (1, 1),
+                    "padding": (0, 0),
+                    "dilation": (1, 1),
+                    **node.kwargs,
+                }
+                output_nhwc = g.call_function(ttnn.max_pool2d, (input_nhwc, in_n, in_h, in_w, in_c, args[1]), kwargs)
+                output_nhwc = g.call_function(ttnn.reshape, (output_nhwc, (out_n, out_h, out_w, out_c)))
+                output = g.call_function(ttnn.permute, (output_nhwc, (0, 3, 1, 2)))
+                user.replace_all_uses_with(output)
+                return None
```

Comment:
This new implementation should transpose and reshape the activation tensors like what we do for convolution (#113), but we are still stuck at the following error message:
```
E       RuntimeError: TT_FATAL @ /home/jdh8/tt-metal/tt_metal/impl/buffers/buffer.cpp:41: page_size % sizeof(uint32_t) == 0
E       info:
E       Page size must be divisible by sizeof(uint32_t) because buffers hold uint32_t values
E       backtrace:
E        --- /home/jdh8/tt-metal/ttnn/ttnn/_ttnn.so(+0x36f2a8) [0x7fdeab2282a8]
E        --- tt::tt_metal::validate_buffer_size_and_page_size(unsigned long, unsigned long, tt::tt_metal::BufferType const&, tt::tt_metal::TensorMemoryLayout const&, std::__1::optional<tt::tt_metal::ShardSpecBuffer> const&)
E        --- tt::tt_metal::Buffer::Buffer(tt::tt_metal::Device*, unsigned long, unsigned long, tt::tt_metal::BufferType, tt::tt_metal::TensorMemoryLayout, std::__1::optional<tt::tt_metal::ShardSpecBuffer> const&, std::__1::optional<bool>, bool)
E        --- /home/jdh8/tt-metal/ttnn/ttnn/_ttnn.so(+0xf05391) [0x7fdeabdbe391]
E        --- tt::tt_metal::tensor_impl::detail::allocate_interleaved_buffer_on_device(unsigned long, tt::tt_metal::Device*, tt::tt_metal::Shape const&, tt::tt_metal::DataType, tt::tt_metal::Layout, tt::tt_metal::MemoryConfig const&)
E        --- tt::tt_metal::tensor_impl::allocate_buffer_on_device(unsigned long, tt::tt_metal::Device*, tt::tt_metal::Shape const&, tt::tt_metal::DataType, tt::tt_metal::Layout, tt::tt_metal::MemoryConfig const&, std::__1::optional<tt::tt_metal::ShardSpecBuffer> const&)
E        --- tt::tt_metal::create_device_tensor(tt::tt_metal::Shape const&, tt::tt_metal::DataType, tt::tt_metal::Layout, tt::tt_metal::Device*, tt::tt_metal::MemoryConfig const&)
E        --- /home/jdh8/tt-metal/ttnn/ttnn/_ttnn.so(_ZN2tt8tt_metal9operation29generic_create_output_tensorsIN4ttnn10operations13data_movement21UntilizeWithUnpaddingEEENS1_21program_output_helperIT_Xsr18has_create_programIS8_EE5valueEE4typeERKS8_RKNSt3__16vectorINS0_6TensorENSD_9allocatorISF_EEEENSD_8optionalINS0_8DataTypeEEENS0_6LayoutERKNSL_INS0_12MemoryConfigEEE+0x1ab) [0x7fdeab4241eb]
E        --- ttnn::operations::data_movement::UntilizeWithUnpadding::create_output_tensors(std::__1::vector<tt::tt_metal::Tensor, std::__1::allocator<tt::tt_metal::Tensor>> const&, std::__1::vector<std::__1::optional<tt::tt_metal::Tensor>, std::__1::allocator<std::__1::optional<tt::tt_metal::Tensor>>> const&) const
E        --- /home/jdh8/tt-metal/ttnn/ttnn/_ttnn.so(+0x572839) [0x7fdeab42b839]
E        --- tt::tt_metal::operation::OldInfraDeviceOperation<std::__1::vector<tt::tt_metal::Tensor, std::__1::allocator<tt::tt_metal::Tensor>>>::tensor_return_value_t ttnn::device_operation::detail::launch_on_single_device<tt::tt_metal::operation::OldInfraDeviceOperation<std::__1::vector<tt::tt_metal::Tensor, std::__1::allocator<tt::tt_metal::Tensor>>>>(unsigned char, tt::tt_metal::operation::OldInfraDeviceOperation<std::__1::vector<tt::tt_metal::Tensor, std::__1::allocator<tt::tt_metal::Tensor>>>::operation_attributes_t const&, tt::tt_metal::operation::OldInfraDeviceOperation<std::__1::vector<tt::tt_metal::Tensor, std::__1::allocator<tt::tt_metal::Tensor>>>::tensor_args_t const&)
E        --- /home/jdh8/tt-metal/ttnn/ttnn/_ttnn.so(+0x37b06a) [0x7fdeab23406a]
E        --- tt::tt_metal::operation::OldInfraDeviceOperation<std::__1::vector<tt::tt_metal::Tensor, std::__1::allocator<tt::tt_metal::Tensor>>>::tensor_return_value_t ttnn::device_operation::detail::invoke<tt::tt_metal::operation::OldInfraDeviceOperation<std::__1::vector<tt::tt_metal::Tensor, std::__1::allocator<tt::tt_metal::Tensor>>>>(unsigned char, tt::tt_metal::operation::OldInfraDeviceOperation<std::__1::vector<tt::tt_metal::Tensor, std::__1::allocator<tt::tt_metal::Tensor>>>::operation_attributes_t const&, tt::tt_metal::operation::OldInfraDeviceOperation<std::__1::vector<tt::tt_metal::Tensor, std::__1::allocator<tt::tt_metal::Tensor>>>::tensor_args_t const&)
E        --- /home/jdh8/tt-metal/ttnn/ttnn/_ttnn.so(+0x37a16b) [0x7fdeab23316b]
```
```
E        --- /home/jdh8/tt-metal/python_env/bin/python3(_PyEval_EvalFrameDefault+0x1f2d) [0x56d31d]
E        --- /home/jdh8/tt-metal/python_env/lib/python3.8/site-packages/torch/lib/libtorch_python.so(+0x8250a6) [0x7fdea927d0a6]

../tt-metal/ttnn/ttnn/decorators.py:326: RuntimeError
```

Reply:
What is the actual shape you try to process?

Reply:
The current test case is relatively simple imo
https://github.com/tenstorrent/pytorch2.0_ttnn/blob/b0e7ff69936be73c5bbe127369a4996dc28c05ad/tests/lowering/pool/test_max_pool_2d.py#L17-L20

Reply:
So you do permute and reshape to 
[1, 64, 64, 3]
then
[1, 64*64*3]
which is 
[1, 12288]

Is that the shape you send to the operation after conversions?

Reply:
No, ignore, I made a mistake here.

The transformation should be
B, C, H, W -> B, H, W, C -> 1, 1, BHW, C